> ### **Assignment 2 - Numpy Array Operations** 
>
> This assignment is part of the course ["Data Analysis with Python: Zero to Pandas"](http://zerotopandas.com). The objective of this assignment is to develop a solid understanding of Numpy array operations. In this assignment you will:
> 
> 1. Pick 5 interesting Numpy array functions by going through the documentation: https://numpy.org/doc/stable/reference/routines.html 
> 2. Run and modify this Jupyter notebook to illustrate their usage (some explanation and 3 examples for each function). Use your imagination to come up with interesting and unique examples.
> 3. Upload this notebook to your Jovian profile using `jovian.commit` and make a submission here: https://jovian.ml/learn/data-analysis-with-python-zero-to-pandas/assignment/assignment-2-numpy-array-operations
> 4. (Optional) Share your notebook online (on Twitter, LinkedIn, Facebook) and on the community forum thread: https://jovian.ml/forum/t/assignment-2-numpy-array-operations-share-your-work/10575 . 
> 5. (Optional) Check out the notebooks [shared by other participants](https://jovian.ml/forum/t/assignment-2-numpy-array-operations-share-your-work/10575) and give feedback & appreciation.
>
> The recommended way to run this notebook is to click the "Run" button at the top of this page, and select "Run on Binder". This will run the notebook on mybinder.org, a free online service for running Jupyter notebooks.
>
> Try to give your notebook a catchy title & subtitle e.g. "All about Numpy array operations", "5 Numpy functions you didn't know you needed", "A beginner's guide to broadcasting in Numpy", "Interesting ways to create Numpy arrays", "Trigonometic functions in Numpy", "How to use Python for Linear Algebra" etc.
>
> **NOTE**: Remove this block of explanation text before submitting or sharing your notebook online - to make it more presentable.


# **Assignment 2 - Numpy Array Operations** 


### Subtitle Here

Write a short introduction about Numpy and list the chosen functions. 

- function 1 = np.mean
- function 2 = np.transpose
- function 3 = np.append
- function 4 = np.swapaxis
- function 5 = np.hsplit

The recommended way to run this notebook is to click the "Run" button at the top of this page, and select "Run on Binder". This will run the notebook on mybinder.org, a free online service for running Jupyter notebooks.

In [1]:
!pip install jovian --upgrade -q

In [14]:
import jovian

<IPython.core.display.Javascript object>

In [15]:
jovian.commit(project='numpy-array-operations')

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "lamoz9ra/numpy-array-operations" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ml/lamoz9ra/numpy-array-operations


'https://jovian.ml/lamoz9ra/numpy-array-operations'

Let's begin by importing Numpy and listing out the functions covered in this notebook.

In [3]:
import numpy as np

## Function 1 - np.mean


In [5]:
a = np.array([[3,4],[5,6],[7,9],[12,13]])

b = np.mean(a)
c = np.mean(a,axis =0)
d = np.mean(a,axis = 1)

print(b)
print(c)
print(d)

7.375
[6.75 8.  ]
[ 3.5  5.5  8.  12.5]


The function is to find mean.

In b we have found the mean of all the values in the arrays. In c, we have set axis = 0, which means that we are finding the means of every first element and the second element in the array separately. i.e 6.75 = 3+5+7+12 / 4 .

In d, we have set axis = 1, which means that we would be finding the mean of every vector in the array. Since we have 4 vectors, we would be having 4 means.

In [6]:
a = np.array([[72,73, 79],[77,96,100],[68,75,77],[96,79,83]])

b = np.mean(a)
c = np.mean(a,axis =0)
d = np.mean(a,axis = 1)



print(b)
print(c)
print(d)

81.25
[78.25 80.75 84.75]
[74.66666667 91.         73.33333333 86.        ]


Explanation about example

In [7]:
# Example 3 - breaking (to illustrate when it breaks)
a = np.array([[43,54, 72],[23,45,10],[68,75,43],[96]])

b = np.mean(a)
c = np.mean(a,axis =0)
d = np.mean(a,axis = 1)

TypeError: unsupported operand type(s) for /: 'list' and 'int'

This breaks because we have in the last vector in the array is a single observation. Whereas all others have 3 observations. Hence mean cannot be calculated.

It is useful to find means of any ordered data such as marks of students.

In [8]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "lamoz9ra/numpy-array-operations" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ml/lamoz9ra/numpy-array-operations


'https://jovian.ml/lamoz9ra/numpy-array-operations'

## Function 2 - np.transpose

This function permutes the dimension of the given array. It returns a view wherever possible. The function takes the following parameters: numpy.transpose(arr, axes), Here arr:The array to be transposed, axes:List of ints, corresponding to the dimensions. By default, the dimensions are reversed

In [9]:
# Example 1 - working
a = np.arange(16).reshape(2,2,4) 

print ('The original array is:' )
print (a)  
print ('\n') 

print('The transposed array is:') 
print (np.transpose(a,(1,0,2)))

The original array is:
[[[ 0  1  2  3]
  [ 4  5  6  7]]

 [[ 8  9 10 11]
  [12 13 14 15]]]


The transposed array is:
[[[ 0  1  2  3]
  [ 8  9 10 11]]

 [[ 4  5  6  7]
  [12 13 14 15]]]


Changing elements in specific positions

In [10]:
# Example 2 - working
a = np.arange(12).reshape(3,4) 

print ('The original array is:' )
print (a)  
print ('\n') 

print('The transposed array is:') 
print ( np.transpose(a))

The original array is:
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]


The transposed array is:
[[ 0  4  8]
 [ 1  5  9]
 [ 2  6 10]
 [ 3  7 11]]


Normal Transpose

In [11]:
# Example 3 - breaking (to illustrate when it breaks)
a = np.arange(12).reshape(3,4) 

print ('The original array is:' )
print (a)  
print ('\n') 

print('The transposed array is:') 
print ( np.transpose(a,0))


The original array is:
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]


The transposed array is:


ValueError: axes don't match array

Explanation about example (why it breaks and how to fix it): Here the program did not execute beacause the axis mentioned are not correct depending on the rows and coloums present we need to adjust the axis. In the above program the axis should be of 2D form.

This function is used to get transform of a array. And also can be used for changes the axis of elements.

In [12]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "lamoz9ra/numpy-array-operations" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ml/lamoz9ra/numpy-array-operations


'https://jovian.ml/lamoz9ra/numpy-array-operations'

## Function 3 - np.append

This function adds values at the end of an input array. The append operation is not inplace, a new array is allocated. Also the dimensions of the input arrays must match otherwise ValueError will be generated.

The function takes the following parameters:numpy.append(arr, values, axis). Here,arr:Input array,values:To be appended to arr. It must be of the same shape as of arr (excluding axis of appending),axis:The axis along which append operation is to be done. If not given, both parameters are flattened.

In [4]:
# Example 1 - working
a = np.array([[1,2,3],[4,5,6]]) 

print ('First array:' )
print (a)
print ('\n')  

print ('Append elements to array:' )
print (np.append(a, [7,8,9])) 

First array:
[[1 2 3]
 [4 5 6]]


Append elements to array:
[1 2 3 4 5 6 7 8 9]


We created a array and later added some more elements to the array.

In [5]:
# Example 2 - working
a = np.array([[1,2,3],[4,5,6]]) 

print ('First array:' )
print (a)
print ('\n') 
print ('Append elements along axis 0:' )
print( np.append(a, [[7,8,9]],axis = 0) )

First array:
[[1 2 3]
 [4 5 6]]


Append elements along axis 0:
[[1 2 3]
 [4 5 6]
 [7 8 9]]


Here we specified axis 0 which means they shouls to added in the first column itself.

In [6]:
# Example 3 - breaking (to illustrate when it breaks)
a = np.array([[1,2,3],[4,5,6]]) 

print ('First array:' )
print (a)
print ('\n') 
print ('Append elements along axis 0:' )
print( np.append(a, [[5,5,5,5]],axis = 0) )

First array:
[[1 2 3]
 [4 5 6]]


Append elements along axis 0:


ValueError: all the input array dimensions for the concatenation axis must match exactly, but along dimension 1, the array at index 0 has size 3 and the array at index 1 has size 4

Explanation about example (why it breaks and how to fix it): Size of the provided new array is out of range so it is difficult to concatenate. To fix this we need to make sure the size of the given array is equal to the original array.

This function can be used to add elements in array when required.

In [7]:
jovian.commit()

NameError: name 'jovian' is not defined

## Function 4 - np.swapaxis

This function interchanges the two axes of an array. A view of the swapped array is returned. The function takes the following parameters:numpy.swapaxes(arr, axis1, axis2), Here arr:Input array whose axes are to be swapped,axis1:An int corresponding to the first axis,axis2:An int corresponding to the second axis.

In [8]:
# Example 1 - working
a = np.arange(8).reshape(2,2,2) 

print ('The original array:') 
print (a) 
print ('\n')  
# now swap numbers between axis 0 (along depth) and axis 2 (along width) 

print ('The array after applying the swapaxes function:' )
print (np.swapaxes(a, 2, 0))

The original array:
[[[0 1]
  [2 3]]

 [[4 5]
  [6 7]]]


The array after applying the swapaxes function:
[[[0 4]
  [2 6]]

 [[1 5]
  [3 7]]]


we interchanged the columns in two different sets

In [9]:
# Example 2 - working
a = np.arange(8).reshape(2,2,2) 

print ('The original array:') 
print (a) 
print ('\n')  
# now swap numbers between axis 0 (along depth) and axis 2 (along width) 

print ('The array after applying the swapaxes function:' )
print (np.swapaxes(a, 1, 2))

The original array:
[[[0 1]
  [2 3]]

 [[4 5]
  [6 7]]]


The array after applying the swapaxes function:
[[[0 2]
  [1 3]]

 [[4 6]
  [5 7]]]


we interchanged the columns in two different sets

In [10]:
# Example 3 - breaking (to illustrate when it breaks)
a = np.arange(8).reshape(2,2,2) 

print ('The original array:') 
print (a) 
print ('\n')  
# now swap numbers between axis 0 (along depth) and axis 2 (along width) 

print ('The array after applying the swapaxes function:' )
print (np.swapaxes(a, 1, 3))

The original array:
[[[0 1]
  [2 3]]

 [[4 5]
  [6 7]]]


The array after applying the swapaxes function:


AxisError: axis2: axis 3 is out of bounds for array of dimension 3

Explanation about example (why it breaks and how to fix it): Here error is because the specifed axis is out of bound. To correct this we can check the boundary of the array and take care of the value when we specify the axis.

This function is used to swap values when required.

In [27]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "aakashns/numpy-array-operations" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ml/aakashns/numpy-array-operations


'https://jovian.ml/aakashns/numpy-array-operations'

## Function 5 - np.hsplit

The numpy.hsplit is a special case of split() function where axis is 1 indicating a horizontal split regardless of the dimension of the input array.

In [11]:
# Example 1 - working
a = np.arange(16).reshape(4,4) 

print ('First array:' )
print (a) 
print ('\n')  

print( 'Horizontal splitting:') 
b = np.hsplit(a,2) 
print(b )

First array:
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]
 [12 13 14 15]]


Horizontal splitting:
[array([[ 0,  1],
       [ 4,  5],
       [ 8,  9],
       [12, 13]]), array([[ 2,  3],
       [ 6,  7],
       [10, 11],
       [14, 15]])]


Here we are spliting the array in 2 different arrays.

In [12]:
# Example 2 - working
a = np.arange(12).reshape(4,3) 

print ('First array:' )
print (a) 
print ('\n')  

print( 'Horizontal splitting:') 
b = np.hsplit(a,3) 
print(b )

First array:
[[ 0  1  2]
 [ 3  4  5]
 [ 6  7  8]
 [ 9 10 11]]


Horizontal splitting:
[array([[0],
       [3],
       [6],
       [9]]), array([[ 1],
       [ 4],
       [ 7],
       [10]]), array([[ 2],
       [ 5],
       [ 8],
       [11]])]


Here we are spliting the array in 3 different arrays

In [13]:
# Example 3 - breaking (to illustrate when it breaks)
a = np.arange(12).reshape(4,3) 

print ('First array:' )
print (a) 
print ('\n')  

print( 'Horizontal splitting:') 
b = np.hsplit(a,5) 
print(b )

First array:
[[ 0  1  2]
 [ 3  4  5]
 [ 6  7  8]
 [ 9 10 11]]


Horizontal splitting:


ValueError: array split does not result in an equal division

Explanation about example (why it breaks and how to fix it): Here the error os because 5 is not a factor of 12 so, dividing the array equally is not possible. So, to fix this we need to take a number which is factor of the size of array.

This function is used to divide the array into equal parts horizontally by a given factor of elements for set.

In [28]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "aakashns/numpy-array-operations" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ml/aakashns/numpy-array-operations


'https://jovian.ml/aakashns/numpy-array-operations'

## Reference Links
Provide links to your references and other interesting articles about Numpy arrays:
* Numpy official tutorial : https://numpy.org/doc/stable/user/quickstart.html
* ...

In [16]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "lamoz9ra/numpy-array-operations" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ml/lamoz9ra/numpy-array-operations


'https://jovian.ml/lamoz9ra/numpy-array-operations'